### PR TITLE
Refactor to be slot based

### PR DESCRIPTION
**Pull request by keithamus**
As part of the work for the new world-if, I am trying to rework component navigation to be more flexible to branding changes. I figured I'd take the opportunity to refactor the code to be much more flexible at accepting new elements inside of the code. Part of that as well is the Accordion...

This PR updates the accordion to use SLOTS! What does this mean? Well, instead of providing objects as the old api:

```js
import React from 'react';
import Accordion from '@economist/component-accordion';
export default (
  <Accordion list={[
    {
      title: 'Sections',
      children: [
        {
          title: 'Latest Updates',
          href: '/latest-updates'
        }
      ]
    },
    {
      title: 'Blogs',
      children: [
        {
          title: 'Buttonwood\'s notebook',
          href: '/blogs/buttonwood',
        }
      ]
    },
  ]} />
);
```

We now have an API which is driven by JSX. It uses almost the same pattern, but is much more extensible in many ways. Here's how the above would be represented with slots:

```js
import React from 'react';
import { Accordion, Panel } from '@economist/component-accordion';
export default (
  <Accordion>
    <Panel header="Sections">
      <ul>
        <li><a href="/latest-updates">Latest Updates</a></li>
      </ul>
    </Panel>
    <Panel header="Blogs">
      <ul>
        <li><a href="/blogs/buttonwood">Buttonwoods notebook</a></li>
      </ul>
    </Panel>
  </Accordion>
);
```

What you lose in a very slight increase in verbosity, you gain in a huge amount more flexibility.

### How does it work?

The way the API works is that `Panel`s are special React components which denote a collapsable section of the accordion for the `Accordion` element to enhance and provide interactivity for. 

### What benefits does it give us?

The real benefits are that the contents of each panel are completely customisable. With the current implementation of the accordion - the contents must be a list of links, however with the new accordion they can be whatever JSX you want to put inside of a panel, for example:

```js
import React from 'react';
import { Accordion, Panel } from '@economist/component-accordion';
export default (
  <Accordion className="birds-list">
    <Panel header="Bluetit">
      <img src="/bird-bluetit.jpg"/>
    </Panel>
    <Panel header="Swallow">
      <img src="/bird-swallow.jpg"/>
    </Panel>
  </Accordion>
);
```

Also, you can have _non collapsable_ parts of the accordion by just adding stuff not inside of a panel element:

```js
import React from 'react';
import { Accordion, Panel } from '@economist/component-accordion';
export default (
  <Accordion className="birds-list">
    <h1>A wonderful list of birds</h1>
    <Panel header="Bluetit">
      <img src="/bird-bluetit.jpg"/>
    </Panel>
    <Panel header="Swallow">
      <img src="/bird-swallow.jpg"/>
    </Panel>
    <small>Terms and conditions apply, this list is non exhaustive</small>
  </Accordion>
);
```

In addition to this, the `header` prop of a panel can be a `node`, so again - anything you want can go in there:

```js
import React from 'react';
import { Accordion, Panel } from '@economist/component-accordion';
export default (
  <Accordion className="birds-list">
    <h1>A wonderful list of birds</h1>
    <Panel header={<h2 className="birds-list__header">Bluetit</h2>}>
      <img src="/bird-bluetit.jpg"/>
    </Panel>
    <Panel header={<h2 className="birds-list__header">Swallow</h2>}>
      <img src="/bird-swallow.jpg"/>
    </Panel>
    <small>Terms and conditions apply, this list is non exhaustive</small>
  </Accordion>
);
```

### What next?

I want to make sure everyone is happy with this changeset. If everyone is, then I'll add tests and polish the commit logs and we can merge. The reason I've made such a big lead up to merging this PR is because _it is a **very** breaking change_!

### CODE REVIEW

**Comment by danielmoll** [comment]
So, if I understand correctly with regards to the css, all the accordion would have is the basics to handle the collapse/expand etc. All other make-it-pretty-stuffs, custom styles, classes etc would be handled in the parent component? That way I guess we get around the problem @umbobabo raised last week.

**Comment by gcedo** [comment]
@keithamus The approach is awesome, and the code is way cleaner and general purpose. It is a breaking change, but IMO it's a breaking change that we need to endure for the sake of future code maintenance and clarity.

**Comment by keithamus** [comment]
> All other make-it-pretty-stuffs, custom styles, classes etc would be handled in the parent component? That way I guess we get around the problem @umbobabo raised last week.

Precisely that. The Accordion component just gives us accordion behaviours - with only enough styling to provide those behaviours (as sadly, CSS is an inevitable part of functionality). You'll notice the CSS is only `display` `block` and `none`, as nothing else needs doing - that styling could also easily be overridden per upstream component, should it require things like animations instead.

In theory this means that this component could also be styled for things like tabs, menus, etc. The main functionality is just "have a set of panels which will only display one opened panel at a time".

File: src/panel.js [hunk 1]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import uniqueId from 'lodash.uniqueid';
+import classNames from 'classnames';
+
+export default function Panel({ className, header, collapse = false, children, onClick, onFocus }) {
+  const idSuffix = uniqueId();
+  const headerId = `accordion__panel-header-${ idSuffix }`;
+  const containerId = `accordion__panel-container-${ idSuffix }`;
+  return (
+    <div className={classNames('accordion__panel', className, { 'accordion__panel--collapsed': collapse })}>
```
Comment (fabiosantoscode) [inline src/panel.js]: Should this be a H* element?

Comment (keithamus) [inline src/panel.js]: The plan was to avoid using H* elements - as they depend on the context - in favour of you passing in your own node. So `<Panel header={<h2>Foo</h2>}>`

Comment (fabiosantoscode) [inline src/panel.js]: Okay, cool

File: src/panel.js [hunk 2]
```diff
@@ -0,0 +1,45 @@
+import React from 'react';
+import uniqueId from 'lodash.uniqueid';
+import classNames from 'classnames';
+
+export default function Panel({ className, header, collapse = false, children, onClick, onFocus }) {
+  const idSuffix = uniqueId();
```
Comment (fabiosantoscode) [inline src/panel.js]: This should be a prop default, as opposed to being generated within. The parent component may not want to generate a counting ID. Take for example fe-blogs, which employs serverside rendering. In this case, performance depends on being able to render the exact same thing on the client and server, so a constant will be passed in instead.

File: src/panel.js [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+        aria-selected={collapse === false}
+      >
+        {header}
+      </div>
+      <div
+        id={containerId}
+        className={classNames('accordion__panel-container', { 'accordion__panel-container--collapsed': collapse })}
+        role="tabpanel"
+        aria-labelledby={headerId}
+        aria-hidden={collapse}
```
Comment (fabiosantoscode) [inline src/panel.js]: Do we want to add the [hidden] attribute?

**Comment by fabiosantoscode** [comment]
Other than my comments, looks like a :rainbow:

File: src/panel.js (hunk 3 above)
Comment (keithamus) [inline src/panel.js]: Not sure. From MDN:

> The hidden attribute must not be used to hide content that could legitimately be shown in another presentation. For example, it is incorrect to use hidden to hide panels in a tabbed dialog, because the tabbed interface is merely a kind of overflow presentation

Comment (fabiosantoscode) [inline src/panel.js]: Since we're using aria-hidden it should be exactly the same.

Comment (keithamus) [inline src/panel.js]: Screen readers can do just that - as a panel gets focus the contents are opened and aria-hidden value is removed, allowing the screen reader to read the list contents. The reason aria-hidden is there in the first place is because the contents are visually hidden (but not _[hidden]_ levels of hidden). In other words, for screen readers and visually, the contents are hidden - but for other presentations they are not (e.g. SEO).

**Comment by keithamus** [comment]
This has been refactored to be a complete solution. Tests have been added. @fabiosantoscode @gcedo @danielmoll can we please get one final review.

**Comment by keithamus** [comment]
Give me 5 mins and I'll get it to 100% coverage.

**Comment by keithamus** [comment]
Tests done. Feel free to review now 😄

**Comment by gcedo** [comment]
LGTM 👍

File: src/panel.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  onFocus,
+  idSuffix = uniqueId(),
+}) {
+  const headerId = `accordion__panel-header-${ idSuffix }`;
+  const containerId = `accordion__panel-container-${ idSuffix }`;
+  return (
+    <div className={classNames('accordion__panel', className, { 'accordion__panel--collapsed': collapse })}>
+      <div
+        id={headerId}
+        className="accordion__panel-header"
```
Comment (umbobabo) [inline src/panel.js]: This will not save us by using specificity if for instance we have 2 different accordion on a page.
Let suppose I use 2 accordions but I want customise all the headers of a specific one. How can I do it without using the  parent selector or the id or replace the Header element itself?

Comment (keithamus) [inline src/panel.js]: You can provide a classname to the panel, and use a selector which reaches to header.

File: src/panel.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+        onClick={onClick}
+        onFocus={onFocus}
+        aria-controls={containerId}
+        aria-selected={collapse === false}
+      >
+        {header}
+      </div>
+      <div
+        id={containerId}
+        className={classNames('accordion__panel-container', { 'accordion__panel-container--collapsed': collapse })}
```
Comment (umbobabo) [inline src/panel.js]: Same problem here, I think.

File: src/panel.js (hunk 4 above)
Comment (umbobabo) [inline src/panel.js]: I thought that having only firs level selector was an important part of this exploration.

**Comment by umbobabo** [comment]
I think that it's really a good solution. I would like anyway to solve the "specificity" issue too, that was at the origin of the discussion, and test it on the fe-blogs context to estimate the impact (in term of css) of this implementation because we could face (due to legacy fault of the re-vamp) some side-effects because the score of the first-level selector will be lower than the score applied by other existing selectors.

**Comment by fabiosantoscode** [comment]
@umbobabo while I also want to strive towards a single-specificity all around, I think this is a good step in that direction.

The thing we have to start doing is probably to create (or find) a `classnames()` BEM wrapper.

```
BEM({ modifier: props.modifier }, 'foo__bar__baz') // -> `foo__bar__baz foo__bar__baz--${props.modifier}`
```

**Comment by keithamus** [comment]
I think we need to strike the right balance between pragmatism and purity. While it would be ideal only ever use one classname for everything, providing the right solution for this would be challenging. I'd sooner say: Live with the idea that occasionally we need to break our single classname rule and have selectors with multiple ones - but mitigate this as much as possible, by writing less coupled code. This is what slots achieves. 

In other words, slots takes us 90% of the way to single class selectors, but the last 10% will take more effort than I think we should deem acceptable. 90% is good enough™

**Comment by umbobabo** [comment]
@keithamus It is a good job, let's play with this on fe-blogs and let see if 90% is good enough also if maybe we could have 100%.
If we need the 100% maybe we could add a prefix prop and propagate down to the children?
